ShardedRedisClient.js

```diff
@@ -0,0 +1,351 @@
+ */
+
+'use strict';
+
+const _ = require('lodash');
+const async = require('async');
+const assert = require('assert');
+const crypto = require('crypto');
+const EventEmitter = require('events').EventEmitter;
+const ShardSet = require('./lib/ShardSet');
+const HostRange = require('./lib/HostRange');
+const WrappedClient = require('./lib/WrappedClient');
+
+const READ_ONLY = [
+  'sismember',
+  'get',
+  'mget',
+  'exists',
+  'hgetall',
+  'llen',
+  'lrange',
+  'scard',
+  'smembers',
+  'srandmember',
+  'sunion', // sunion assumes that each key is on the same shard
+  'zrevrange',
+  'zcard',
+  'zcount',
+  'zscore',
+  'zrange',
+  'zrangebyscore'
+];
+
+const SHARDABLE = [
+  'sadd',
+  'expire',
+  'ttl',
+  'sismember',
+  'srem',
+  'get',
+  'mget',
+  'exists',
+  'scard',
+  'smembers',
+  'sunion', // sunion assumes that each key is on the same shard
+  'hdel',
+  'hget',
+  'hincrby',
+  'hset',
+  'hmset',
+  'hgetall',
+  'llen',
+  'lpush',
+  'lrange',
+  'ltrim',
+  'set',
+  'setnx',
+  'setex',
+  'psetex',
+  'del',
+  'srandmember',
+  'zrevrange',
+  'incr',
+  'zadd',
+  'zcard',
+  'zcount',
+  'zrem',
+  'zscore',
+  'zrange',
+  'zrangebyscore',
+  'zremrangebyrank',
+  'zremrangebyscore'
+];
+
+function getClientIndex(key, wrappedClients) {
+  const hash = crypto.createHash('sha1').update(String(key)).digest('hex');
+  const hashCut = hash.substring(0, 4);
+  const hashNum = parseInt(hashCut, 16);
+
+  return hashNum % wrappedClients.length;
+}
+
+function getReadClient(clientIndex, wrappedClients, readSlave) {
+  const wrappedClient = wrappedClients[clientIndex];
+  const slaveOk = readSlave || wrappedClient.readPreference === 'slave';
+
+  return slaveOk ? wrappedClient.getSlave() : wrappedClient.get();
+}
+
+function getWriteClient(clientIndex, wrappedClients) {
+  const wrappedClient = wrappedClients[clientIndex];
+
+  return wrappedClient.get();
+}
+
+function findMasterClient(key, wrappedClients) {
+  const clientIndex = getClientIndex(key, wrappedClients);
+  const wrappedClient = wrappedClients[clientIndex];
+
+  return wrappedClient.get();
+}
+
+function getWrappedClient(key, wrappedClients) {
+  const clientIndex = getClientIndex(key, wrappedClients);
+
+  return wrappedClients[clientIndex];
+}
+
+function findMatchedClient(key, cmd, wrappedClients, readSlave) {
+  const clientIndex = getClientIndex(key, wrappedClients);
+  const isReadCmd = READ_ONLY.indexOf(cmd) >= 0;
+
+  return isReadCmd ? getReadClient(clientIndex, wrappedClients, readSlave) : getWriteClient(clientIndex, wrappedClients);
+}
+
+class ShardedRedisClient extends EventEmitter {
+  
+  constructor(configurationArray, options) {
+    assert(Array.isArray(configurationArray), 'first argument \'configurationArray\' must be an array.');
+
+    if (typeof options === 'boolean')
+      options = { usePing: options };
+    else if (!options)
+      options = {};
+
+    if (options.usePing !== false)
+      options.usePing = true;
```

Comment:
why are we allowing multiple types for options above?

also if `options.usePing === 0`, you'll set `options.usePing` to `true`?